File: 17_gpsanta.cpp
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-
// -- BEGIN LICENSE BLOCK ----------------------------------------------
// -- END LICENSE BLOCK ------------------------------------------------
//----------------------------------------------------------------------
/*!\file
 *
 * \author  Christoph Rist <rist.christoph@gmail.com>
 * \date    2016-12-28
 *
 */
//----------------------------------------------------------------------
/***********************************
 * IEEE SB Passau Adventskalender
 * Problem 17
 */

#include <memory>
#include <iostream>
#include <iomanip>
#include <vector>
#include <set>
#include <map>
#include <stack>
#include <cmath>
#include <algorithm>

//#define TEST
#ifdef TEST
#include "io_redirect.h"
#endif

using namespace std;


class Graph
{
public:
  using Node = pair<uint8_t, uint8_t>;

  void addEdge(const Node& first_node, const Node& second_node)
  {
    addNode(first_node);
    addNode(second_node);
    nodes_[first_node].insert(second_node);
  }

  uint32_t nbSimplePaths(const Node& start, const Node& end)
  {
    set<Node> visited{};
    uint32_t nb_paths = 0;

    stack<pair<Node, set<Node>::const_iterator>> s;
    s.emplace(start, nodes_[start].cbegin());

    while(!s.empty())
    {
      auto& n = s.top();

      if (n.first == end || n.second == nodes_[n.first].cend())
      {
        if (n.first == end)
          ++nb_paths;
        visited.erase(n.first);
        s.pop();
      }
      else
      {
        const Node& v = *n.second;
        ++n.second;

        if (visited.find(v) == visited.cend())
        {
          s.emplace(v, nodes_.at(v).cbegin());
          visited.insert(v);
        }
      }

    }

    return nb_paths;
  }
private:
  void addNode(const Node& n)
  {
    if (nodes_.find(n) == nodes_.cend())
    {
      nodes_.emplace(n, set<Node>{});
    }
  }

  map<Node, set<Node>> nodes_;
};

Graph buildGraph(uint8_t size)
{
  Graph g{};
  for (uint8_t x = 1; x < size; ++x)
  {
    for (uint8_t y = 0; y < x; ++y)
    {
      g.addEdge({x - 1, y}, {x, y});
      g.addEdge({x, y}, {x, y + 1});
    }
  }
  return g;
}

int main(int argc, char* argv[])
{
#ifdef TEST
  //////////////////////// INPUT/OUTPUT //////////////////////////
  if (!redirect_io(argc, argv))
    return 0;
#endif

  uint32_t nb_testcases;
  cin >> nb_testcases;
  vector<uint8_t> sizes{};
  // parse
  string l;
  getline(cin, l);
  for (uint32_t i = 0; i < nb_testcases; ++i)
  {
    getline(cin, l);
    const auto it = std::find(l.cbegin(), l.cend(), 'x');
    if (it == l.cend())
      throw runtime_error("Invalid input");

    const int32_t first = stoi(string(l.cbegin(), it));
    const int32_t second = stoi(string(it + 1, l.cend()));

    if (first < 0 || first != second || first > 255)
      throw runtime_error("Invalid input");

    sizes.push_back(static_cast<uint8_t>(first));
  }

  for (uint8_t s : sizes)
  {
    auto g = buildGraph(s + 1);
    cout <<g.nbSimplePaths({0, 0}, {s, s}) <<endl;
  }


#ifdef TEST
  cleanup_io();
#endif

  return 0;
}

File: 17_gpsanta.py
#!/usr/bin/env python3
from sys import maxsize
import sys
from itertools import combinations


__author__ = "Christoph Rist"
__license__ = "MIT"

DEBUG = False


class Graph:
    def __init__(self):
        self.nodes = set()
        self.edges = {}
        self.nodes = {}

    def add_node(self, node):
        if node not in self.nodes:
            self.nodes[node] = []

    def set_edge(self, start, end, weight):
        self.add_node(start)
        self.add_node(end)

        if (start, end) not in self.edges:
            self.nodes[start].append(end)
        self.edges[start, end] = weight

    def set_undirected_edge(self, first_node, second_node, weight=1):
        self.set_edge(first_node, second_node, weight)
        self.set_edge(second_node, first_node, weight)

    def shortest_route(self, start, end):
        """ Return shortest path between start and end.

        This is a dijkstra implementation.
        """
        if start not in self.nodes or end not in self.nodes:
            return None

        previous = {}
        unvisited = set(self.nodes.keys())
        distance = {n: maxsize for n in unvisited}

        distance[start] = 0

        while unvisited:
            u = min(unvisited, key=distance.get)
            if distance[u] == maxsize:
                return None
            if u == end:
                break
            unvisited.remove(u)

            for neighbor in self.nodes[u]:
                if neighbor not in unvisited:
                    continue

                d = distance[u] + self.edges[u, neighbor]
                if neighbor not in distance or d < distance[neighbor]:
                    distance[neighbor] = d
                    previous[neighbor] = u

        sequence = [end]
        u = end
        while u in previous:
            u = previous[u]
            sequence.append(u)
        sequence.reverse()
        return sequence, distance[end]

    def diameter(self):
        # find shortest path for every pair of nodes
        # currently assume an undirected graph!
        max_len = 0
        for pair in combinations(self.nodes, 2):
            seq = self.shortest_route(pair[0], pair[1])
            if seq is None:
                break
            if len(seq) - 1 > max_len:
                max_len = len(seq) - 1
        else:
            return max_len
        return None

    def nb_simple_paths(self, start, end):
        """ Calculate number of simple path between start and end

        This is a depth-first search
        """
        if start not in self.nodes or end not in self.nodes:
            return None

        visited = set()
        nb_paths = 0

        def search(node, goal):
            nonlocal nb_paths
            visited.add(node)

            if node == goal:
                nb_paths += 1
            else:
                for neighbor in self.nodes[node]:
                    if neighbor not in visited:
                        search(neighbor, goal)

            visited.remove(node)

        search(start, end)
        return nb_paths


def build_graph(size):
    graph = Graph()
    # every non-diagonal x-y-coordinate
    for x in range(1, size):
        for y in range(x):
            graph.set_undirected_edge((x, y), (x - 1, y))
            graph.set_undirected_edge((x, y), (x, y + 1))

    return graph


def main():

    if DEBUG:
        sys.stdin = open("samples/17.1_input.txt")

    n = int(input())
    sizes = []
    for _ in range(n):
        # format is INT x INT, parse this:
        size = list(map(int, [x.strip() for x in input().split("x")]))
        # sanity check
        if len(size) != 2 or size[0] != size[1]:
            raise RuntimeError("invalid input")
        sizes.append(size[0])

    if max(sizes) > 12:
        raise RuntimeError()

    for size in sizes:
        # this seems to be a matter of definition
        if size == 1:
            print(1)
            continue

        graph = build_graph(size)
        print(graph.nb_simple_paths((0, 0), (size - 1, size - 1)) * 2)


if __name__ == "__main__":
    main()
